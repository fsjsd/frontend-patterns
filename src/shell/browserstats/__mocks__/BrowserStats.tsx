import React from 'react'

const BrowserStats = () => {
  return (
    <div>BrowserStats Mock</div>
  )
}

export default BrowserStats