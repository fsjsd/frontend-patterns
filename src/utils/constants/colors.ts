
export const TRAFFIC_LIGHTS = {
  GREEN: "#2daf2d",
  AMBER: "#a86323",
  RED: "#972020",
}