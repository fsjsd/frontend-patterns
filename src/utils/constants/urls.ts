export const GITHUB_ROOT = "https://github.com/fsjsd/frontend-patterns/tree/main/src"