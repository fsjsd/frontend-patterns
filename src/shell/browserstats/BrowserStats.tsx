import React, { useEffect, useReducer, useRef } from 'react'
import { TRAFFIC_LIGHTS } from '../../utils/constants/colors';

// adapted from : https://github.com/tibotiber/react-fps-stats/blob/development/src/index.js

export function fpsRating(currentFps: number): import("csstype").Property.Color {
  if (currentFps >= 60) {
    return TRAFFIC_LIGHTS.GREEN;
  }
  if (currentFps >= 50) {
    return TRAFFIC_LIGHTS.AMBER;
  }
  return TRAFFIC_LIGHTS.RED;
}

const BrowserStats = () => {
  const graphWidth = 70;

  const [state, dispatch] = useReducer(
    state => {
      const currentTime = Date.now()
      if (currentTime > state.prevTime + 1000) {
        const nextFPS = Math.round(
          (state.frames * 1000) / (currentTime - state.prevTime)
        )
        return {
          max: Math.max(state.max, nextFPS),
          len: Math.min(state.len + 1, graphWidth),
          fps: [...state.fps, nextFPS].slice(-graphWidth),
          frames: 1,
          prevTime: currentTime
        }
      } else {
        return { ...state, frames: state.frames + 1 }
      }
    },
    {
      len: 0,
      max: 0,
      frames: 0,
      prevTime: Date.now(),
      fps: []
    }
  )

  const requestRef = useRef<number>()
  const tick = () => {
    dispatch()
    requestRef.current = requestAnimationFrame(tick)
  }

  useEffect(() => {
    requestRef.current = requestAnimationFrame(tick)
    return () => {
      requestRef.current && cancelAnimationFrame(requestRef.current)
    }
  }, []);

  const { fps, len } = state //max, 
  const currentFps = fps[len - 1];

  return (
    <div>
      {currentFps && <b aria-label='FPS' style={{ color: fpsRating(currentFps) }}>{currentFps}fps</b>}
    </div>
  )
}

export default BrowserStats

