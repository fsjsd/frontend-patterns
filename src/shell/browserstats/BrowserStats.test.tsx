import React from 'react'
import { render, waitFor, screen } from "@testing-library/react";
import BrowserStats, { fpsRating } from "./BrowserStats"
import { TRAFFIC_LIGHTS } from '../../utils/constants/colors';

const RealDate = Date;

describe('BrowserStats', () => {
  let startTime: number;
  beforeEach(() => {
    // 1555928430000 = new Date('2019-04-22T10:20:30Z').getTime()
    startTime = 1555928430000;
    global.Date.now = jest.fn(() => startTime);
  });

  afterEach(() => {
    global.Date = RealDate;
  });

  it("renders and matches snapshot", async () => {
    const rafSpy = jest.spyOn(window, 'requestAnimationFrame');
    rafSpy.mockImplementation(
      (callback: FrameRequestCallback): number => {
        setTimeout(() => {
          callback(0);
          startTime += 1000; // add second to start time
        }, 1);
        return 0;
      });
    const { container } = render(<BrowserStats />)
    await waitFor(() => expect(screen.queryByLabelText("FPS")).toBeInTheDocument());
    expect(container).toMatchSnapshot();
    rafSpy.mockRestore();
  });

  it.each([
    [0, TRAFFIC_LIGHTS.RED],
    [10, TRAFFIC_LIGHTS.RED],
    [50, TRAFFIC_LIGHTS.AMBER],
    [52, TRAFFIC_LIGHTS.AMBER],
    [60, TRAFFIC_LIGHTS.GREEN],
    [62, TRAFFIC_LIGHTS.GREEN]
  ])('shows %s color for %s value', async (fps, expectedColor) => {
    expect(fpsRating(fps)).toBe(expectedColor);
  })
});