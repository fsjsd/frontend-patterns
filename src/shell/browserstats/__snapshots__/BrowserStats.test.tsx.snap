// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrowserStats renders and matches snapshot 1`] = `
<div>
  <div>
    <b
      aria-label="FPS"
      style="color: rgb(151, 32, 32);"
    >
      1
      fps
    </b>
  </div>
</div>
`;
